tsl/src/nodes/decompress_chunk/exec.c

```diff
@@ -500,139 +498,203 @@ perform_vectorized_aggregation(DecompressChunkState *chunk_state)
+				ExecProcNode(linitial(chunk_state->csstate.custom_ps));
+			if (TupIsNull(compressed_slot))
+			{
+				/* All compressed batches are processed. */
+				break;
+			}
 
-		/* We have at least one value */
-		decompressed_scan_slot->tts_isnull[0] = false;
+			/* Decompress data */
+			bool isnull;
+			Datum value =
+				slot_getattr(compressed_slot, column_description->compressed_scan_attno, &isnull);
+			if (isnull)
+			{
+				continue; /* TODO: Or maybe error out? */
+			}
 
-		// TODO: Handle segment by values
-		CompressedDataHeader *header = (CompressedDataHeader *) PG_DETOAST_DATUM(value);
-		ArrowArray *arrow = NULL;
+			/* We have at least one value */
+			decompressed_scan_slot->tts_isnull[0] = false;
 
-		if (chunk_state->bulk_decompression_context == NULL)
-		{
-			chunk_state->bulk_decompression_context =
-				AllocSetContextCreate(MemoryContextGetParent(batch_state->per_batch_context),
-									  "bulk decompression",
-									  /* minContextSize = */ 0,
-									  /* initBlockSize = */ 64 * 1024,
-									  /* maxBlockSize = */ 64 * 1024);
-		}
+			// TODO: Handle segment by values
+			CompressedDataHeader *header = (CompressedDataHeader *) PG_DETOAST_DATUM(value);
+			ArrowArray *arrow = NULL;
 
-		DecompressAllFunction decompress_all =
-			tsl_get_decompress_all_function(header->compression_algorithm);
-		Assert(decompress_all != NULL);
+			DecompressAllFunction decompress_all =
+				tsl_get_decompress_all_function(header->compression_algorithm);
+			Assert(decompress_all != NULL);
 
-		MemoryContext context_before_decompression =
-			MemoryContextSwitchTo(chunk_state->bulk_decompression_context);
+			MemoryContext context_before_decompression =
+				MemoryContextSwitchTo(chunk_state->bulk_decompression_context);
 
-		/* Fixme, type for column is wrong because its partial(sum(int4)) */
-		column_description->typid = 23;
+			/* Fixme, type for column is wrong because its partial(sum(int4)) */
+			column_description->typid = 23;
 
-		arrow = decompress_all(PointerGetDatum(header),
-							   column_description->typid,
-							   batch_state->per_batch_context);
+			arrow = decompress_all(PointerGetDatum(header),
+								   column_description->typid,
+								   batch_state->per_batch_context);
 
-		Assert(arrow != NULL);
+			Assert(arrow != NULL);
 
-		MemoryContextReset(chunk_state->bulk_decompression_context);
-		MemoryContextSwitchTo(context_before_decompression);
+			MemoryContextReset(chunk_state->bulk_decompression_context);
+			MemoryContextSwitchTo(context_before_decompression);
 
 #if defined(HAVE__BUILTIN_OP_OVERFLOW) && defined(USE_CUSTOM_SUM_IMPL)
 #define SUM_LANES 4
-		uint32 lane_result[] = { 0, 0, 0, 0 };
-		uint32 lane_buffer[] = { 0, 0, 0, 0 };
-		bool lane_overflow[] = { false, false, false, false };
+			uint32 lane_result[] = { 0, 0, 0, 0 };
+			uint32 lane_buffer[] = { 0, 0, 0, 0 };
+			bool lane_overflow[] = { false, false, false, false };
 
-		int cur_lane = 0;
+			int cur_lane = 0;
```

Comment:
You can exploit the fact that the arrow arrays are padded at the end, to always work in full lanes.